FLEDGE.md

```diff
@@ -437,12 +441,54 @@ The bundle may be fetched using credentials like cookies, as described in the [s
 
 For the JSON response, only the `https` scheme is supported -- the `uuid-in-package` scheme isn't supported as that scheme doesn't support CORS.
 
+#### 2.5.2 Using Response Headers
+
+An alternative way to pass DirectFromSellerSignals without subresource bundles is via the `Ad-Auction-Signals` response header of some `fetch()` request, together with the `directFromSellerSignalsHeaderAdSlot` parameter on `navigator.runAdAuction()`.
+
+With this method, a [`fetch()`](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) call is made by some script on the page (this call may be made in a subframe), with an extra option, `{adAuctionHeaders: true}`:
+
+```javascript
+let fetchResponse = await fetch("https://seller.com/signals", {adAuctionHeaders: true});
+```
+
+The browser will make the request as in a normal (optionless) fetch call, with the exception that the request will also include a request header, `Sec-Ad-Auction-Fetch: ?1`. This header indicates to the server that any `Ad-Auction-Signals` response header from the server will only be loaded in auctions via `directFromSellerSignalsHeaderAdSlot` (this is analogous to the guarantees of `X-FLEDGE-Auction-Only` and `Sec-Fetch-Dest: webbundle` from the [subresource bundle version](#251-using-subresource-bundles) -- scripts on the page cannot set the `Sec-Ad-Auction-Fetch: ?1` request header without using the `{adAuctionHeaders: true}` option).
+
+The value of the Ad-Auction-Signals header should be JSON formatted, with the following schema:
+
+```json
+Ad-Auction-Signals=[{
+  "adSlot": "adSlot/1",
+  "sellerSignals": {/*...*/},
+  "auctionSignals": {/*...*/},
+  "perBuyerSignals": {"https://buyer1.example": /*...*/, /*...*/}
+},
+{
+  "adSlot": "adSlot/2",
+  "sellerSignals": {/*...*/},
+  "auctionSignals": {/*...*/},
+  "perBuyerSignals": {"https://buyer1.example": /*...*/, /*...*/}
+},
+/*...*/
+]
+```
+
+When invoking `navigator.runAdAuction()`, `directFromSellerSignalsHeaderAdSlot` is used to lookup the signals intended for that auction. `directFromSellerSignalsHeaderAdSlot` is a string that should match the `adSlot` value contained in some `Ad-Auction-Signals` response served from the origin of that auction's seller. Note that for multi-seller or component auctions, each component auction can specify its own `directFromSellerSignalsHeaderAdSlot`, and the response should be served from that component auction's seller. Different sellers may safely use the same `adSlot` names without conflict. If `directFromSellerSignalsHeaderAdSlot` matches multiple `adSlot`s from header responses, one of those `adSlot` responses will be chosen arbitrarily.
```

Comment:
>If `directFromSellerSignalsHeaderAdSlot` matches multiple `adSlot`s from header responses, one of those `adSlot` responses will be chosen arbitrarily.

Why not LIFO versus arbitrary?

Reply:
I could look into switching to LIFO if there's interest -- the current WIP MVP implementation stores an unordered set of responses, but it could possibly be switched to a queue, maybe even in a later version (switching to a non-arbitrary order from an arbitrary order shouldn't be a breaking change).

Reply:
+1 to switching to a LIFO data structure. Ad techs may choose to define the `adSlot` as an element on the page containing an ad that may refresh the ad over time. It seems unlikely to want the same exact set of signals when a newer response with the same `adSlot` is available on the page.

Similarly, have you considered how resource management should work in this flow? Once the ad auction is completed, is there any way to dispose the stale signals, or is this an unavoidable source of memory leaks? We plan on iterating on the amount of data transferable within this header and starting small; do you anticipate any issues when approaching O(kilobytes) of JSON?

Reply:
OK, I think we can use a LIFO order [0], where a fetch for adSlot "foo" for fetch / seller origin seller.com would overwrite any previous stored signals for the (seller.com, "foo") key tuple.

In the initial version [1], all server responses are stored until the top-level page is unloaded, and there is no way to otherwise remove old responses from memory; each new `fetch()` with `{adAuctionHeaders: true}` that captures new headers will result in additional memory consumption, and we also parse potentially all of those responses in arbitrary order, only stopping if a matching adSlot is found.

Currently, the entire set of responses is re-parsed and searched for each top-level, component auction, or non-component auction on the page using the new directFromSellerSignals. So, if there are many responses, or they are large, this will add to the auction latency.

With the new LIFO order, the signals would only be parsed once, and further lookups would just be map lookups (unless more signals have arrived).

Individual responses greater than 1000 bytes are currently discarded [2]. We could consider increasing this, but headers are generally small, and several server implementations and CDNs impose their own limits. Also, the signals get copied from the browser process to worklet processes.

Do folks think these proposed changes would address all your concerns?

[0] Internally in the Chromium implementation, it would technically be a FIFO queue of the JSON string responses -- we'd process all of them during auctions, overwriting the value signals for each (fetch / seller origin, adSlot string) key tuple as we process the list, then empty the processed FIFO -- this would mean that more recent responses overwrite older responses, or LIFO semantics. The spec already works in a similar manner (except that JSON parsing is synchronous in spec language, so the parsing happens in the FETCH spec patch). The reason for processing the JSON during auctions is to avoid complexity around potential races / waiting in auctions until processing has completed if started in the network interceptor.

**Note**, this means that the memory for old signals wouldn't be freed until an auction is run in this new proposal.

[1] The initial version was cherry-picked into Chrome Beta 117.0.5938.35, but we'll need approval on an I2S to be able to flip the controlling flag on (currently working on getting approval, which requires we complete the spec).

[2] https://source.chromium.org/chromium/chromium/src/+/main:content/browser/interest_group/ad_auction_url_loader_interceptor.cc;l=166;drc=6f4c64436342c818aa41e6a5c55034e74ec9c6b6

Reply:
Thanks Caleb, the LIFO order optimization sounds great!

Reply:
Here's a PR to change this for the explainer (the spec already matches this behavior): https://github.com/WICG/turtledove/pull/890